### PR TITLE
Accelerometer calibration on reconnect with flight controller.

### PR DESCRIPTION
**Pull request by aftersomemath**

### CODE REVIEW

File: include/iarc7_fc_comms/CommonFcComms.hpp [hunk 1]
```diff
@@ -124,16 +134,20 @@ namespace FcComms{
...
         // Service to arm copter
         ros::ServiceServer uav_arm_service;
 
         iarc7_msgs::OrientationThrottleStamped::ConstPtr last_direction_command_message_ptr_;
 
+        iarc7_msgs::LandingGearContactsStamped::ConstPtr last_contact_switch_message_ptr_;
+
         bool have_new_direction_command_message_ = false;
 
+        bool have_contact_switch_message_ = false;
```
Comment (amiller27) [inline include/iarc7_fc_comms/CommonFcComms.hpp]: Don't need this, just test if `last_contact_switch_message_ptr_` is null.  We should also be keeping track of not only if we have a message, but if we have a recent one.

File: include/iarc7_fc_comms/CommonFcComms.hpp [hunk 2]
```diff
@@ -97,6 +101,12 @@ namespace FcComms{
             have_new_direction_command_message_ = true;
         }
 
+        inline void contactSwitchMessageHandler(
+                const iarc7_msgs::LandingGearContactsStamped::ConstPtr& message) {
+            last_contact_switch_message_ptr_ = message;
```
Comment (amiller27) [inline include/iarc7_fc_comms/CommonFcComms.hpp:106]: Are we sure it's safe to assign a `shared_ptr` like this?  I really don't know

File: include/iarc7_fc_comms/CommonFcComms.hpp [hunk 3]
```diff
@@ -478,13 +518,39 @@ void CommonFcComms<T>::reconnect() {
     if(status == FcCommsReturns::kReturnOk)
     {
         ROS_INFO("iarc7_fc_comms: Succesful reconnection to flight controller");
+        calibrateAccelerometer();
```
Comment (amiller27) [inline include/iarc7_fc_comms/CommonFcComms.hpp:536]: Do we not care about handling failures here?

File: include/iarc7_fc_comms/CommonFcComms.hpp (hunk 2 above)
Comment (aftersomemath) [reply include/iarc7_fc_comms/CommonFcComms.hpp:106]: I don't either now that you mention it. We have been doing it other places as well

File: include/iarc7_fc_comms/CommonFcComms.hpp (hunk 1 above)
Comment (aftersomemath) [reply include/iarc7_fc_comms/CommonFcComms.hpp]: Fair

File: include/iarc7_fc_comms/CommonFcComms.hpp (hunk 3 above)
Comment (aftersomemath) [reply include/iarc7_fc_comms/CommonFcComms.hpp:536]: This function handles its failures. Since we don't use exceptions there isn't anything to check

Comment (amiller27) [reply include/iarc7_fc_comms/CommonFcComms.hpp:536]: I see

File: include/iarc7_fc_comms/CommonFcComms.hpp (hunk 2 above)
Comment (amiller27) [reply include/iarc7_fc_comms/CommonFcComms.hpp:106]: A little research shows this should be right, it's being copy assigned, which releases the pointer previously held by `last_contact_switch_message_ptr_` and increases the reference count on what it's being assigned to.

Comment (aftersomemath) [reply include/iarc7_fc_comms/CommonFcComms.hpp:106]: cool

File: include/iarc7_fc_comms/CommonFcComms.hpp [hunk 4]
```diff
@@ -236,7 +248,32 @@ FcCommsReturns CommonFcComms<T>::init()
...
+           && ros::Time::now()
+              < start_time
+                + ros::Duration(CommonConf::kContactSwitchStartupTimeout)) {
+        ros::spinOnce();
+        ros::Duration(0.005).sleep();
+    }
+
+    if (last_contact_switch_message_ptr_ == nullptr)
+    {
+        ROS_WARN("Contact switch message not received within the startup timeout");
```
Comment (amiller27) [inline include/iarc7_fc_comms/CommonFcComms.hpp]: We should return `FcCommsReturns::kReturnError` in this case

Comment (aftersomemath) [reply include/iarc7_fc_comms/CommonFcComms.hpp]: I designed this to support working with or without the contact switches. This is important for this working for test programs that won't launch the foot switches and don't need too. I don't think we do because this isn't an error. Just a warning.

File: include/iarc7_fc_comms/CommonFcComms.hpp (hunk 1 above)
Comment (aftersomemath) [reply include/iarc7_fc_comms/CommonFcComms.hpp]: done

File: include/iarc7_fc_comms/CommonFcComms.hpp (hunk 4 above)
Comment (amiller27) [reply include/iarc7_fc_comms/CommonFcComms.hpp]: Then it should be a parameter.  We shouldn't be ignoring the switches on the physical drone

Comment (aftersomemath) [reply include/iarc7_fc_comms/CommonFcComms.hpp]: That's definitely more engineering work than this is worth. It requires updating everything that uses fc comms because its an interface change.

The velocity publisher will yell if the foot switches aren't working.

It's not necessary for fc comms operation to have the foot switches.

Comment (amiller27) [reply include/iarc7_fc_comms/CommonFcComms.hpp]: I meant a `rosparam` parameter

Comment (aftersomemath) [reply include/iarc7_fc_comms/CommonFcComms.hpp]: I thought of that too. Still an interface change because users will need to set the parameter.

Comment (amiller27) [reply include/iarc7_fc_comms/CommonFcComms.hpp]: Sure, it's a interface change.  That's because we added a feature.  If you really don't want to break compatibility, make it default to not check the switches.  In reality, fixing the interface isn't that complicated.

Comment (aftersomemath) [reply include/iarc7_fc_comms/CommonFcComms.hpp]: I agree, but there's no advantage in the short term in fixing this. Saying we should fix this is the same as saying lots of other things have the same priority to be fixed.

Comment (amiller27) [reply include/iarc7_fc_comms/CommonFcComms.hpp]: Repeatedly saying exactly what you just said is why all of FcComms's parameters are compiled into the source instead of loaded from rosparam like they should be

Comment (aftersomemath) [reply include/iarc7_fc_comms/CommonFcComms.hpp]: I still have no problem with that decision for the reasons I gave.

Comment (aftersomemath) [reply include/iarc7_fc_comms/CommonFcComms.hpp]: Done

File: include/iarc7_fc_comms/CommonFcComms.hpp [hunk 5]
```diff
@@ -255,19 +268,23 @@ FcCommsReturns CommonFcComms<T>::init()
...
+        }
+
+        if (last_contact_switch_message_ptr_ == nullptr)
+        {
+            ROS_ERROR("Contact switch message not received within the startup timeout");
+            return FcCommsReturns::kReturnError;
+        }
     }
     else
     {
```
Comment (amiller27) [inline include/iarc7_fc_comms/CommonFcComms.hpp]: Can we get rid of this `else` block?

**Review by amiller27 — APPROVED** [review]
